pinot-controller/src/main/java/com/linkedin/pinot/controller/api/storage/PinotSegmentUploadUtils.java

```diff
@@ -0,0 +1,167 @@
+    PinotFS pinotFS;
+    URI srcUri;
+    URI dstUri;
+    try {
+      srcUri = new URI(segmentUploaderConfig.getHeaders().getHeaderString(FileUploadDownloadClient.CustomHeaders.DOWNLOAD_URI));
+      pinotFS = PinotFSFactory.getInstance().init(_controllerConf, srcUri);
+      dstUri = constructFinalLocation(segmentMetadata, segmentVersion);
+      pinotFS.copy(srcUri, dstUri);
+    } catch (Exception e) {
+      LOGGER.error("Could not copy file to final directory");
+      throw new RuntimeException(e);
+    }
+
+    HttpHeaders headers = segmentUploaderConfig.getHeaders();
+    String tableName = segmentMetadata.getTableName();
+    String segmentName = segmentMetadata.getName();
+    String tableNameWithType = getTableNameWithType(segmentName, tableName);
+
+    ZNRecord znRecord = _pinotHelixResourceManager.getSegmentMetadataZnRecord(tableNameWithType, segmentName);
+
+    // Brand new segment, not refresh, directly add the segment
+    if (znRecord == null) {
+      String downloadUrl = constructDownloadUrl(tableName, segmentName, provider, pinotFS, dstUri, segmentVersion);
+      _pinotHelixResourceManager.addNewSegment(segmentMetadata, downloadUrl);
+      return;
+    }
+
+    // Segment already exists, refresh if necessary
+    OfflineSegmentZKMetadata existingSegmentZKMetadata = new OfflineSegmentZKMetadata(znRecord);
+    long existingCrc = existingSegmentZKMetadata.getCrc();
+
+    try {
+      // Modify the custom map in segment ZK metadata
+      String segmentZKMetadataCustomMapModifierStr = headers.getHeaderString(FileUploadDownloadClient.CustomHeaders.SEGMENT_ZK_METADATA_CUSTOM_MAP_MODIFIER);
+      SegmentZKMetadataCustomMapModifier segmentZKMetadataCustomMapModifier;
+      if (segmentZKMetadataCustomMapModifierStr != null) {
+        segmentZKMetadataCustomMapModifier =
+            new SegmentZKMetadataCustomMapModifier(segmentZKMetadataCustomMapModifierStr);
+      } else {
+        // By default, use REPLACE modify mode
+        segmentZKMetadataCustomMapModifier =
+            new SegmentZKMetadataCustomMapModifier(SegmentZKMetadataCustomMapModifier.ModifyMode.REPLACE, null);
+      }
+      existingSegmentZKMetadata.setCustomMap(
+          segmentZKMetadataCustomMapModifier.modifyMap(existingSegmentZKMetadata.getCustomMap()));
+
+      // Update ZK metadata and refresh the segment if necessary
+      long newCrc = Long.valueOf(segmentMetadata.getCrc());
+      if (newCrc == existingCrc) {
+        // New segment is the same as the existing one, only update ZK metadata without refresh the segment
+        if (!_pinotHelixResourceManager.updateZkMetadata(existingSegmentZKMetadata)) {
+          throw new RuntimeException(
+              "Failed to update ZK metadata for segment: " + segmentName + " of table: " + tableNameWithType);
+        }
+      } else {
+        _pinotHelixResourceManager.refreshSegment(segmentMetadata, existingSegmentZKMetadata, dstUri.getPath());
+      }
+    } catch (Exception e) {
```

Comment:
Please catch specific exception where possible and let runtime exceptions get thrown as would happen automatically. Not sure there is much value to catching e and throwing e without further logging/processing.